lib/compiler/test/beam_validator_SUITE.erl

```diff
@@ -808,6 +810,90 @@ mrt_1(Bool) ->
+              [{label,1},
+               {func_info,{atom,bs_saved_position_units},{atom,ctx_test_8},1},
+               {label,2},
+               {'%',
+                   {var_info,
+                       {x,0},
+                       [{type,{t_bs_context,8,0,0}},accepts_match_context]}},
+               {move,nil,{x,0}},
+               return]},
+          {function,no_errors,1,4,
+              [{label,3},
+               {func_info,{atom,bs_saved_position_units},{atom,no_errors},1},
+               {label,4},
+               {'%',{var_info,{x,0},[accepts_match_context]}},
+               {test,bs_start_match3,{f,3},1,[{x,0}],{x,1}},
+               {bs_get_position,{x,1},{x,0},2},
+               {test,bs_test_unit,{f,5},[{x,1},8]},
+               {bs_set_position,{x,1},{x,0}},
+               {test,bs_get_binary2,
+                   {f,5},
+                   2,
+                   [{x,1},{atom,all},1,{field_flags,[unsigned,big]}],
+                   {x,2}},
+               {bs_set_position,{x,1},{x,0}},
+               {bs_get_tail,{x,1},{x,0},3},
+               {test,is_eq_exact,{f,5},[{x,2},{x,0}]},
+               {move,{x,1},{x,0}},
+               %% Context unit should be 8 here.
+               {call_only,1,{f,2}},
+               {label,5},
+               {bs_get_tail,{x,1},{x,0},2},
+               {jump,{f,3}}]},
+          {function,some_errors,1,7,
+              [{label,6},
+               {func_info,{atom,bs_saved_position_units},{atom,some_errors},1},
+               {label,7},
+               {'%',{var_info,{x,0},[accepts_match_context]}},
+               {test,bs_start_match3,{f,6},1,[{x,0}],{x,1}},
+               {bs_get_position,{x,1},{x,0},2},
+               {test,bs_get_binary2,
+                   {f,8},
+                   2,
+                   [{x,1},{atom,all},4,{field_flags,[unsigned,big]}],
+                   {x,2}},
+               {bs_set_position,{x,1},{x,0}},
+               {test,bs_test_unit,{f,9},[{x,1},3]},
+               {bs_set_position,{x,1},{x,0}},
+               {bs_get_tail,{x,1},{x,0},3},
+               {test,is_eq_exact,{f,8},[{x,2},{x,0}]},
+               {move,{x,1},{x,0}},
+               %% Context unit should be 12 here, failing validation.
+               {call_only,1,{f,2}},
+               {label,8},
+               {bs_get_tail,{x,1},{x,0},2},
+               {jump,{f,6}},
+               {label,9},
+               %% Context unit should be 4 here.
+               {move,nil,{x,0}},
+               return]}],
+         10},
+
+    Errors = beam_val(M),
+
+    [{{bs_saved_position_units,some_errors,1},
+      {{call_only,1,{f,2}},
+       14,
+       {bad_arg_type,{x,0},
+                     {t_bs_context,12,0,0},
+                     {t_bs_context,8,0,0}}}}] = Errors,
+
+    [<<0,1,2,3,4>>, <<5,6,7,8,9>>] = bspu_1(id(<<0,1,2,3,4,5,6,7,8,9>>)),
+    [] = bspu_1(id(<<>>)),
+
+    ok.
+
+bspu_1(<<Bin/binary>> = Bin) ->
+    [Chunk || <<Chunk:5/binary>> <= Bin].
```

Comment:
I suggest moving this part of the test case to `bs_match_SUITE` so that the code will be tested also without optimization and compiled for OTP 21.

Reply:
Fixed.